### ISSUE TITLE
bndrun editor: Dragging bundle from Browse Repos to Run Bundles / Run Requirements should use stricter version-range instead of open range version='1.2.3'

### ISSUE THREAD

**Issue by chrisrueger** [bndtools/bnd#5816]
At EclipseCon2023 @juergen-albert pointed out in his presentation that in OSGI 

version='1.2.3' basically means "this version and everything above". Doing this could be dangerous,as it might pull in higher versions you do not expect.

See https://docs.osgi.org/specification/osgi.core/7.0.0/framework.module.html#i2655136
<img width="1301" alt="image" src="https://github.com/bndtools/bnd/assets/188422/e4ce5180-3ff7-442a-a68c-14efd72d838a">

## Problem

When I drag a bundle in the bndrun-editor to e.g. the Run Bundles then it uses the '1.2.3' version instead of a range.

<img width="927" alt="image" src="https://github.com/bndtools/bnd/assets/188422/f99bd45d-1ae7-4157-8c9a-ef3e8888f448">

**Actual result:**
`org.apache.commons.collections4;version='4.1.0'`

**Expected / suggested result:**
`org.apache.commons.collections4;version='[4.1.0,4.1.1)'` or `org.apache.commons.collections4;version='[4.1.0,4.1.0]'`

Having a stricter and smaller range could prevent problems and unexpected events.

Edit: Just want to add, that I think this is not a bug. But since bndtools should "do good" things, I think the current behavir might lead to unexpected events. So consider this a "Usability improvement".

**Comment by pkriens** [comment]
a) the idea of the resolver is to only supply the functional bundles, the bundles you care about to be in your application. It is hard to imagine your customer cares about apache collections? The idea of the resolver is that you specify the bundles the customer is interested, they tend to be leaf bundles that nobody depends on. The resolver than drags in the rest.

b) I am not a fan of these small ranges until you get into problems. The more freedom the resolver has, the faster and better solutions it finds. This is not maven where you want a hard reference to a a specific JAR.

c) I could live with restricting it to the next major range but definitely not by default to the next micro, that is way too restrictive and makes life downstream too complicated.

**Comment by chrisrueger** [comment]
> c) I could live with restricting it to the next major range but definitely not by default to the next micro, that is way too restrictive and makes life downstream too complicated.

👍

I think this was my intention... I just didn't write it😃 my bad. 
My point was, that i was surprised that version=1.2.3 is not an exact match. 
Usually we use a range until the next major when editing by hand, but i just noticed when dragging in the ui.

On that note: i think the runquires window also creates a version='1.2.3' when using drag and drop. So maybe the same might apply there.

**Comment by chrisrueger** [comment]
When you use the Resolver, then version ranges are inserted e.g. `org.apache.commons.collections4;version='[4.1.0,4.1.1)',\` 

So maybe one would expect similar behaviour when something "gets inserted" by "the tool" (regardless of a user by "drag & drop" or the resolver).

**Comment by pkriens** [comment]
I think that is different. That version is chosen by the resolver, it would even be better to have some identity. The solution very narrowly depends on having exactly the right version.

**Comment by chrisrueger** [comment]
I have create PR proposal: https://github.com/bndtools/bnd/pull/5823

I basically implemented your suggestion 

> c) I could live with restricting it to the next major range

**Comment by kriegfrj** [comment]
> Are your talking about `runrequires` or `runbundles`? @chrisrueger case is only for the `runbundles` and I have to say, I'm with him on that. If I decide to pull in Bundle manually and drag it in by the Version, then that is exactly the Bundle Version I want and not any other Version of it. The resolver also produces a Bundle with a strict Version range for the runbundles and the result should be the same here.

I think this discussion needs a little more nuance as there is more than one way that you can drag a bundle into `-runrequires` or `-runbundles`.

The **Browse Repos** view is not a simple list, but a tree. The root nodes of the tree are the bundles sorted by repository and then by bsn (so if the same bundle bsn is in two different repos, it will appear as two root nodes). Under each root node are one or more leaf nodes representing the versions of that bsn that are available in that repo. Current behaviour is different depending on whether you are dragging a root node or a leaf node:

#### -runrequires
* Dragging a root node adds a `bnd.identity` requirement on the bsn only, with no version attribute.
* Dragging a leaf node adds a `bnd.identity requirement with a version attribute equal to the full precision of `major.minor.micro.qualifier` (which I think corresponds to a minimum requirement of that precise version, not an exact requirement for that precise version).
* No de-duplication is done - the UI will happily let you add the same bsn multiple times 

#### -runbundles
* Dragging a root nodes adds the unversioned bsn
* Dragging a leaf node adds the bsn with a version attribute equal to the `major.minor.micro` (the qualifier is dropped). This is again a *minimum* requirement rather than a *precise* requirement.
* Duplicate bsns are overwritten with the new version.

Behaviour is also different if you drag bundles from the Workspace repo: it will add `version=latest` regardless of whether you drag the leaf or the root. I believe that it should use `version=snapshot` rather than `version=latest` in this case.

#### resolver
* For bundles that come from the workspace repo, the resolver uses `version=snapshot`.
* For all other bundles, the resolver always adds version attributes that are precise to the depth of the *micro* portion of the version number (ie, any bundle with the same major.minor.micro version will be acceptable, regardless of the value of any qualifier).

#### Analysis

Firstly, I think we are free to change the default behaviour if we think it is appropriate, because it is more part of the UI rather than the build. Users immediately get GUI feedback of what the result of their UI action has done, and if it is not to their liking they have the ability to edit the source to fix it to what they want. So the question is, do we need to change this default behaviour to something more appropriate?

I think that @pkriens is right when it comes to `-runrequires` version ranges are preferable. But I think that the current behaviour of specifying an open range is too flexible - I think being open to the next major range is more appropriate.

But I think that @juergen-albert is also correct: when dragging into `-runbundles` (as opposed to `-runrequires`), then we are bypassing or overriding the resolver's results. So @pkriens's comments about not wanting to restrict the resolver's freedom don't apply. Manually manipulating `-runbundles` in this way is not usually done without a very specific reason, and so typically the more precise version specification would be more appropriate. If the user needs a broader range, they always have the option of hand-editing.

We also have the option of making the default behaviour configurable via the preferences dialog. This is what we did for version range specification for the quickfix processor (#4474). This is not particularly onerous and (if @pkriens still wishes to less precise ranges for dragging behaviour for `-runbundles` in spite of the above clarification) it would give us a way to keep everyone happy.

#### Proposal
* Change `-runbundles` dragging behaviour when dragging from workspace repository to use `version=snapshot` rather than `version=latest`.
* For dragging root nodes, use unversioned bsn (ie, maintain current behaviour).
* For dragging leaf nodes: 
   * Change `-runrequires` to use range to the next major version (what @chrisrueger has done in his PR, but for `-runrequires` rather than `runbundles`.
   * Change `-runbundles` to use range to the next micro version (mimicking the resolver behaviour).

Thoughts?

**Comment by chrisrueger** [comment]
Thanks @kriegfrj for this detailed analysis. I think this goes much beyond what I initially had in mind (I basically was only talking about `dragging leaf nodes of non-workspace bundles to runbundles` ) :) But nevertheless it is great feedback.

Since this broadens the scope of this initial issue description, the question is: Should we now solve all the nuances? Or should we focus on `runbundles` first and do the rest in a separate issue / PR?

> default behaviour configurable via the preferences dialog

Personally I think a configuration preference might be a bit too much here. Because "dragging a leaf node (a specific) version" to `runbundles` should in in just one way and hopefully in a way which is most helpful for the user with least (negative) surprise. 

About your proposals:

> * Change `-runbundles` dragging behaviour when dragging from workspace repository to use `version=snapshot` rather than `version=latest`.

I can also look into that. Seems to be the same method which currently has:

```
if (RepoUtils.isWorkspaceRepo(bundleVersion.getParentBundle()
			.getRepo()))
			attribs.put(Constants.VERSION_ATTRIBUTE, VERSION_LATEST);
```

> For dragging root nodes, use unversioned bsn (ie, maintain current behaviour).

Oh, I haven't tested this (didn't even know that this is possible. I always clicked on the specific version. I will check that. 

> For dragging leaf nodes:
> Change -runrequires to use range to the next major version (what @chrisrueger has done in his PR, but for -runrequires rather than runbundles.

This is where I am unsure since I do not have enough knowledge. I would vote for doing this in another PR, since `runrequires` was not something I initially had in mind, when opening the issue. I am not sure `runrequires` even has a problem in the context of this issue. 

> Change -runbundles to use range to the next micro version (mimicking the resolver behaviour).

This is what I initially had in mind. 
I will commit a change for this which will do a `bumpMicro()` instead of `bumpMajor()` (but **only** `runbundles` ).

We can then check how this looks.

**Comment by chrisrueger** [comment]
@kriegfrj I pushed a change https://github.com/chrisrueger/bnd/commit/710621b60b354018358e3988f49cfe18405c8c56 which does this:

> Dragging a leaf node adds the bsn with a version attribute equal to the major.minor.micro (the qualifier is dropped). This is again a minimum requirement rather than a precise requirement.

I have not yet done this:

> Change -runbundles dragging behaviour when dragging from workspace repository to use version=snapshot rather than version=latest.

This is quickly done but I would wait for confirmation.

**Comment by chrisrueger** [comment]
> I believe that it should use `version=snapshot` rather than `version=latest` in this case.

Could you explain what is the difference? I remember @juergen-albert explained it to me once, but I can't remember. 
I will be happy to add this to the manual later.

**Comment by juergen-albert** [comment]
@chrisrueger `version=snapshot` means the Version build in your workspace. `version=latest` is the highest version found. Usually the workspace holds the latest Version. It is possible though, that you may have a newer Version of one of the projects in your workspace in one of your repositories. 

It's rare, but possible.

**Comment by juergen-albert** [comment]
@kriegfrj 

> Change -runbundles dragging behaviour when dragging from workspace repository to use version=snapshot rather than version=latest.

+1

>For dragging root nodes, use unversioned bsn (ie, maintain current behaviour).

+1

>For dragging leaf nodes:
>Change -runrequires to use range to the next major version (what @chrisrueger has done in his PR, but for -runrequires >rather than runbundles.

Here I'm not sure and is most likely a thing of personal preference. If I drag a leaf into runrequires, I explicitly dragged in a Version. If I want it more open, then I can modify it afterwards. For people that are new to that, It might be nice to have some tooling support for the modification. But this is something for the future. (Just as an idea: Double click on the requires entry can open a window like PDE has, where the upper and lower bound can be configured.)

>Change -runbundles to use range to the next micro version (mimicking the resolver behaviour).

+1

**Comment by chrisrueger** [comment]
@juergen-albert thanks for the clarification.
I added that info as a comment so that we have it somewhere. We should also find a place in the documentation where this could be added.

https://github.com/chrisrueger/bnd/commit/763a74300f354945436097f08c15b750aa4511b3

I added a comment to summarize the current status of the PR https://github.com/bndtools/bnd/pull/5823#issuecomment-1778916280

**Comment by kriegfrj** [comment]
> I will commit a change for this which will do a `bumpMicro()` instead of `bumpMajor()` (but **only** `runbundles` ).
> 
> We can then check how this looks.

:+1: to be clear, I know what I wrote went beyond the original scope of your proposal,  and I wasn't suggesting you be lumped with the responsibility of implementing it all other than the part that this PR relates to. I just thought it was good to take a step back and make sure our change is consistent across other parts of the UI, which is why I thought it was worth taking a step back to look at the bigger picture. :smile: Thank you for your efforts.

**Comment by kriegfrj** [comment]
> > Change -runrequires to use range to the next major version (what @chrisrueger has done in his PR, but for -runrequires >rather than runbundles.
> 
> Here I'm not sure and is most likely a thing of personal preference. If I drag a leaf into runrequires, I explicitly dragged in a Version. If I want it more open, then I can modify it afterwards. For people that are new to that, It might be nice to have some tooling support for the modification. But this is something for the future.

Is the current behaviour an exact match though, or a lower bound only? I had thought it was a lower bound only, so my proposal actually narrows the range by giving it an upper bound?

> (Just as an idea: Double click on the requires entry can open a window like PDE has, where the upper and lower bound can be configured.)
> 
:+1:

**Comment by juergen-albert** [comment]
yes, sorry. You are correct. I was under the impression that it would come with an upper bound, but it does not. 

So, I stand correct and change my response to a +1 :-)

**Comment by chrisrueger** [comment]
Ok since @pkriens merged https://github.com/bndtools/bnd/pull/5823 into master, I think we can close this issue and move the `runrequires` discussion to a new issue.

**Comment by pkriens** [comment]
We can keep this one open for the -runrequires discussion.

**Comment by pkriens** [comment]
I am ok to support up to major version when the version line is dragged. This is the standard OSGi approach that something is compatible to the next major release. For the generic bsn name, I think we should have no version.

**Comment by chrisrueger** [comment]
> I am ok to support up to major version when the version line is dragged. This is the standard OSGi approach that something is compatible to the next major release. For the generic bsn name, I think we should have no version.

Thanks @pkriens I am currently working on the `-runrequires` part of this issue. I think I'll have an update soon.

**Questions:**

Can you confirm (correct me) if the following `runrequires` results after dragging or adding are ok?

**1. Adding / dragging a Workspace Bundle**
Expected Result: `bnd.identity;id='com.my.bundle';version=snapshot`

**2. Adding / dragging a repo bundle-version (leaf node)**
Expected Result: `bnd.identity;id='org.apache.felix.logback';version:Version='[1.0.2,2.0.0)'`

**3. Adding / dragging a repo bundle (root node without version)**
Expected Result: `bnd.identity;id='org.apache.felix.logback'`

Note: the `bnd.identity;` above is based on the _simplified requirements_ as configured in preferences:
<img width="644" alt="image" src="https://github.com/bndtools/bnd/assets/188422/4eba5e3d-0371-4eb9-a9c8-5f68ef40c80d">

To summarize `-runrequires` vs `-runbundles` in case of non-workspace bundles:

`-runrequires` will use a [version,major) e.g. `[1.2.3,2.0.0)` range
`-runbundles` will use a smaller range [version,micro) e.g. `[1.2.3,1.2.4)` range
